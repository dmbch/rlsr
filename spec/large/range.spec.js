/* eslint-env node, jest */
const R = require('ramda');
const diff = require('recursive-diff').getDiff;
const transform = require('../../src/transform');
const getEnv = require('../fixtures/env-large.fixture');
const getMessage = require('../fixtures/message-large.fixture');

const applyMainScope = R.assoc('scope', 'four');
const applyMinorLevel = R.assoc('type', 'feat');
const applyOtherScope = R.assoc('scope', 'three');
const applyMultiScope = R.assoc('body', 'affects: two, three');
const applyMajorLevel = R.assoc('footer', 'BREAKING CHANGE: text ');

describe('large test: exact mode', () => {
  Date.now = jest.fn(() => 0);

  it('applies patch message', () => {
    // make the date function returd deterministic values

    const env = getEnv([applyMainScope(getMessage('foo'))]);
    env.config.mode = 'range';

    const exp = transform(env);

    expect(exp.packages.one.dependencies.four).toBe('^4.5.0'); // not changed
    expect(exp.packages.one.version).toBe('1.2.3');
    expect(exp.packages.three.dependencies.four).toBe('^4');
    expect(exp.packages.three.version).toBe('3.4.6');
    expect(exp.packages.two.dependencies.three).toBe('2.3.4 - 3'); // also in range
    expect(exp.packages.two.version).toBe('2.3.4');

    expect(diff(env, exp)).toMatchSnapshot();
  });

  it('applies minor message', () => {
    const env = getEnv([applyMainScope(applyMinorLevel(getMessage('bar')))]);
    env.config.mode = 'range';
    const exp = transform(env);

    expect(exp.packages.one.dependencies.four).toBe('^4.5.0'); // not changed
    expect(exp.packages.one.version).toBe('1.2.3');
    expect(exp.packages.three.dependencies.four).toBe('^4');
    expect(exp.packages.three.version).toBe('3.4.6');
    expect(exp.packages.two.dependencies.three).toBe('2.3.4 - 3'); // also in range
    expect(exp.packages.two.version).toBe('2.3.4');

    expect(diff(env, exp)).toMatchSnapshot();
  });

  it('applies major message', () => {
    const env = getEnv([applyMainScope(applyMajorLevel(getMessage('baz')))]);
    env.config.mode = 'range';
    const exp = transform(env);

    expect(exp.packages.one.dependencies.four).toBe('^5'); // changed
    expect(exp.packages.one.version).toBe('1.2.4');
    expect(exp.packages.three.dependencies.four).toBe('^5');
    expect(exp.packages.three.version).toBe('3.4.6');
    expect(exp.packages.two.dependencies.three).toBe('2.3.4 - 3'); // in range
    expect(exp.packages.two.version).toBe('2.3.4');

    expect(diff(env, exp)).toMatchSnapshot();
  });

  it('applies several messages', () => {
    const env = getEnv([
      applyMinorLevel(applyMainScope(getMessage('taz'))), // four 4.6.0
      getMessage('raz'), // one 1.2.4
      applyOtherScope(getMessage('roo')), // three 3.4.6
    ]);
    env.config.mode = 'range';

    const exp = transform(env);

    expect(exp.packages.one.dependencies.four).toBe('^4.5.0'); // in range
    expect(exp.packages.one.version).toBe('1.2.4');
    expect(exp.packages.three.dependencies.four).toBe('^4');
    expect(exp.packages.three.version).toBe('3.4.6');
    expect(exp.packages.two.dependencies.three).toBe('2.3.4 - 3'); // in range
    expect(exp.packages.two.version).toBe('2.3.4');

    expect(diff(env, exp)).toMatchSnapshot();
  });

  it('applies messages with multiple affected packages', () => {
    const env = getEnv([
      applyMinorLevel(getMessage('waz')), // one 1.3.0
      applyMultiScope(getMessage('tar')), // two 2.3.5 three 3.4.6
    ]);
    env.config.mode = 'range';

    const exp = transform(env);

    expect(exp.packages.one.version).toBe('1.3.0');
    expect(exp.packages.three.version).toBe('3.4.6');
    expect(exp.packages.two.dependencies.three).toBe('2.3.4 - 3');
    expect(exp.packages.two.version).toBe('2.3.5');

    expect(diff(env, exp)).toMatchSnapshot();
  });

  it('applies minor then major', () => {
    const env = getEnv([applyMinorLevel(getMessage('raz'))]); // one 1.3.0
    env.config.mode = 'range';

    const exp1 = transform(env);
    expect(diff(env, exp1)).toMatchSnapshot();

    exp1.messages = [applyOtherScope(applyMajorLevel(getMessage('rar')))]; // three 4.0.0

    const exp2 = transform(exp1);
    expect(diff(exp1, exp2)).toMatchSnapshot();
  });

  it('applies major then minor', () => {
    const env = getEnv([applyMajorLevel(getMessage('raz'))]);
    env.config.mode = 'range';

    const exp1 = transform(env);
    expect(diff(env, exp1)).toMatchSnapshot();

    exp1.messages = [applyOtherScope(applyMinorLevel(getMessage('rar')))];

    const exp2 = transform(exp1);
    expect(diff(exp1, exp2)).toMatchSnapshot();
  });
});
