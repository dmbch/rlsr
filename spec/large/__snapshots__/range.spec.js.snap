// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`large test: exact mode applies major message 1`] = `
Object {
  "appRoot": "/",
  "changelog": Object {
    "2.0.0": Object {
      "changes": Object {
        "four": Object {
          "determinedIncrementLevel": "major",
          "messages": Array [
            Object {
              "affected": Array [
                "four",
              ],
              "body": "BREAKING CHANGE: text ",
              "subject": "baz",
              "type": "fix",
            },
          ],
          "version": "5.0.0",
        },
      },
      "date": "Thu, 01 Jan 1970 00:00:00 GMT",
    },
  },
  "config": Object {
    "mode": "range",
  },
  "consts": Object {
    "levels": Array [
      "patch",
      "minor",
      "major",
    ],
    "nsp": "rlsr",
    "rlsrLatest": "rlsr-latest",
  },
  "dbg": [Function],
  "log": [Function],
  "mainPackage": Object {
    "rlsr": Object {
      "determinedIncrementLevel": 2,
    },
    "version": "2.0.0",
  },
  "messages": Array [],
  "packages": Array [
    Object {
      "dependencies": Object {
        "five": "^4.5.0",
        "four": "^4.5.0",
      },
      "name": "one",
      "version": "1.2.3",
    },
    Object {
      "dependencies": Object {
        "three": "2.3.4 - 3",
      },
      "name": "two",
      "version": "2.3.4",
    },
    Object {
      "dependencies": Object {
        "four": "4.0.0 - 4.5.6",
      },
      "name": "three",
      "version": "3.4.5",
    },
    Object {
      "name": "four",
      "version": "5.0.0",
    },
    Object {
      "name": "five",
      "version": "4.5.6",
    },
  ],
}
`;

exports[`large test: exact mode applies major then minor 1`] = `
Object {
  "appRoot": "/",
  "changelog": Object {
    "2.0.0": Object {
      "changes": Object {
        "one": Object {
          "determinedIncrementLevel": "major",
          "messages": Array [
            Object {
              "affected": Array [
                "one",
              ],
              "body": "BREAKING CHANGE: text ",
              "subject": "raz",
              "type": "fix",
            },
          ],
          "version": "2.0.0",
        },
      },
      "date": "Thu, 01 Jan 1970 00:00:00 GMT",
    },
  },
  "config": Object {
    "mode": "range",
  },
  "consts": Object {
    "levels": Array [
      "patch",
      "minor",
      "major",
    ],
    "nsp": "rlsr",
    "rlsrLatest": "rlsr-latest",
  },
  "dbg": [Function],
  "log": [Function],
  "mainPackage": Object {
    "rlsr": Object {
      "determinedIncrementLevel": 2,
    },
    "version": "2.0.0",
  },
  "messages": Array [],
  "packages": Array [
    Object {
      "dependencies": Object {
        "five": "^4.5.0",
        "four": "^4.5.0",
      },
      "name": "one",
      "version": "2.0.0",
    },
    Object {
      "dependencies": Object {
        "three": "2.3.4 - 3",
      },
      "name": "two",
      "version": "2.3.4",
    },
    Object {
      "dependencies": Object {
        "four": "4.0.0 - 4.5.6",
      },
      "name": "three",
      "version": "3.4.5",
    },
    Object {
      "name": "four",
      "version": "4.5.6",
    },
    Object {
      "name": "five",
      "version": "4.5.6",
    },
  ],
}
`;

exports[`large test: exact mode applies major then minor 2`] = `
Object {
  "appRoot": "/",
  "changelog": Object {
    "2.0.0": Object {
      "changes": Object {
        "one": Object {
          "determinedIncrementLevel": "major",
          "messages": Array [
            Object {
              "affected": Array [
                "one",
              ],
              "body": "BREAKING CHANGE: text ",
              "subject": "raz",
              "type": "fix",
            },
          ],
          "version": "2.0.0",
        },
      },
      "date": "Thu, 01 Jan 1970 00:00:00 GMT",
    },
  },
  "config": Object {
    "mode": "range",
  },
  "consts": Object {
    "levels": Array [
      "patch",
      "minor",
      "major",
    ],
    "nsp": "rlsr",
    "rlsrLatest": "rlsr-latest",
  },
  "dbg": [Function],
  "log": [Function],
  "mainPackage": Object {
    "rlsr": Object {
      "determinedIncrementLevel": -1,
    },
    "version": "2.0.0",
  },
  "messages": Array [],
  "packages": Array [
    Object {
      "dependencies": Object {
        "five": "^4.5.0",
        "four": "^4.5.0",
      },
      "name": "one",
      "version": "2.0.0",
    },
    Object {
      "dependencies": Object {
        "three": "2.3.4 - 3",
      },
      "name": "two",
      "version": "2.3.4",
    },
    Object {
      "dependencies": Object {
        "four": "4.0.0 - 4.5.6",
      },
      "name": "three",
      "version": "3.4.5",
    },
    Object {
      "name": "four",
      "version": "4.5.6",
    },
    Object {
      "name": "five",
      "version": "4.5.6",
    },
  ],
}
`;

exports[`large test: exact mode applies messages with multiple affected packages 1`] = `
Object {
  "appRoot": "/",
  "changelog": Object {
    "1.9.0": Object {
      "changes": Object {
        "one": Object {
          "determinedIncrementLevel": "minor",
          "messages": Array [
            Object {
              "affected": Array [
                "one",
              ],
              "body": "",
              "subject": "waz",
              "type": "feat",
            },
          ],
          "version": "1.3.0",
        },
        "three": Object {
          "determinedIncrementLevel": "patch",
          "messages": Array [
            Object {
              "affected": Array [
                "two",
                "three",
              ],
              "body": "
",
              "subject": "tar",
              "type": "fix",
            },
          ],
          "version": "3.4.6",
        },
        "two": Object {
          "determinedIncrementLevel": "patch",
          "messages": Array [
            Object {
              "affected": Array [
                "two",
                "three",
              ],
              "body": "
",
              "subject": "tar",
              "type": "fix",
            },
          ],
          "version": "2.3.5",
        },
      },
      "date": "Thu, 01 Jan 1970 00:00:00 GMT",
    },
  },
  "config": Object {
    "mode": "range",
  },
  "consts": Object {
    "levels": Array [
      "patch",
      "minor",
      "major",
    ],
    "nsp": "rlsr",
    "rlsrLatest": "rlsr-latest",
  },
  "dbg": [Function],
  "log": [Function],
  "mainPackage": Object {
    "rlsr": Object {
      "determinedIncrementLevel": 1,
    },
    "version": "1.9.0",
  },
  "messages": Array [],
  "packages": Array [
    Object {
      "dependencies": Object {
        "five": "^4.5.0",
        "four": "^4.5.0",
      },
      "name": "one",
      "version": "1.3.0",
    },
    Object {
      "dependencies": Object {
        "three": "2.3.4 - 3",
      },
      "name": "two",
      "version": "2.3.5",
    },
    Object {
      "dependencies": Object {
        "four": "4.0.0 - 4.5.6",
      },
      "name": "three",
      "version": "3.4.6",
    },
    Object {
      "name": "four",
      "version": "4.5.6",
    },
    Object {
      "name": "five",
      "version": "4.5.6",
    },
  ],
}
`;

exports[`large test: exact mode applies minor message 1`] = `
Object {
  "appRoot": "/",
  "changelog": Object {
    "1.9.0": Object {
      "changes": Object {
        "four": Object {
          "determinedIncrementLevel": "minor",
          "messages": Array [
            Object {
              "affected": Array [
                "four",
              ],
              "body": "",
              "subject": "bar",
              "type": "feat",
            },
          ],
          "version": "4.6.0",
        },
      },
      "date": "Thu, 01 Jan 1970 00:00:00 GMT",
    },
  },
  "config": Object {
    "mode": "range",
  },
  "consts": Object {
    "levels": Array [
      "patch",
      "minor",
      "major",
    ],
    "nsp": "rlsr",
    "rlsrLatest": "rlsr-latest",
  },
  "dbg": [Function],
  "log": [Function],
  "mainPackage": Object {
    "rlsr": Object {
      "determinedIncrementLevel": 1,
    },
    "version": "1.9.0",
  },
  "messages": Array [],
  "packages": Array [
    Object {
      "dependencies": Object {
        "five": "^4.5.0",
        "four": "^4.5.0",
      },
      "name": "one",
      "version": "1.2.3",
    },
    Object {
      "dependencies": Object {
        "three": "2.3.4 - 3",
      },
      "name": "two",
      "version": "2.3.4",
    },
    Object {
      "dependencies": Object {
        "four": "4.0.0 - 4.5.6",
      },
      "name": "three",
      "version": "3.4.5",
    },
    Object {
      "name": "four",
      "version": "4.6.0",
    },
    Object {
      "name": "five",
      "version": "4.5.6",
    },
  ],
}
`;

exports[`large test: exact mode applies minor then major 1`] = `
Object {
  "appRoot": "/",
  "changelog": Object {
    "1.9.0": Object {
      "changes": Object {
        "one": Object {
          "determinedIncrementLevel": "minor",
          "messages": Array [
            Object {
              "affected": Array [
                "one",
              ],
              "body": "",
              "subject": "raz",
              "type": "feat",
            },
          ],
          "version": "1.3.0",
        },
      },
      "date": "Thu, 01 Jan 1970 00:00:00 GMT",
    },
  },
  "config": Object {
    "mode": "range",
  },
  "consts": Object {
    "levels": Array [
      "patch",
      "minor",
      "major",
    ],
    "nsp": "rlsr",
    "rlsrLatest": "rlsr-latest",
  },
  "dbg": [Function],
  "log": [Function],
  "mainPackage": Object {
    "rlsr": Object {
      "determinedIncrementLevel": 1,
    },
    "version": "1.9.0",
  },
  "messages": Array [],
  "packages": Array [
    Object {
      "dependencies": Object {
        "five": "^4.5.0",
        "four": "^4.5.0",
      },
      "name": "one",
      "version": "1.3.0",
    },
    Object {
      "dependencies": Object {
        "three": "2.3.4 - 3",
      },
      "name": "two",
      "version": "2.3.4",
    },
    Object {
      "dependencies": Object {
        "four": "4.0.0 - 4.5.6",
      },
      "name": "three",
      "version": "3.4.5",
    },
    Object {
      "name": "four",
      "version": "4.5.6",
    },
    Object {
      "name": "five",
      "version": "4.5.6",
    },
  ],
}
`;

exports[`large test: exact mode applies minor then major 2`] = `
Object {
  "appRoot": "/",
  "changelog": Object {
    "1.9.0": Object {
      "changes": Object {
        "one": Object {
          "determinedIncrementLevel": "minor",
          "messages": Array [
            Object {
              "affected": Array [
                "one",
              ],
              "body": "",
              "subject": "raz",
              "type": "feat",
            },
          ],
          "version": "1.3.0",
        },
      },
      "date": "Thu, 01 Jan 1970 00:00:00 GMT",
    },
  },
  "config": Object {
    "mode": "range",
  },
  "consts": Object {
    "levels": Array [
      "patch",
      "minor",
      "major",
    ],
    "nsp": "rlsr",
    "rlsrLatest": "rlsr-latest",
  },
  "dbg": [Function],
  "log": [Function],
  "mainPackage": Object {
    "rlsr": Object {
      "determinedIncrementLevel": -1,
    },
    "version": "1.9.0",
  },
  "messages": Array [],
  "packages": Array [
    Object {
      "dependencies": Object {
        "five": "^4.5.0",
        "four": "^4.5.0",
      },
      "name": "one",
      "version": "1.3.0",
    },
    Object {
      "dependencies": Object {
        "three": "2.3.4 - 3",
      },
      "name": "two",
      "version": "2.3.4",
    },
    Object {
      "dependencies": Object {
        "four": "4.0.0 - 4.5.6",
      },
      "name": "three",
      "version": "3.4.5",
    },
    Object {
      "name": "four",
      "version": "4.5.6",
    },
    Object {
      "name": "five",
      "version": "4.5.6",
    },
  ],
}
`;

exports[`large test: exact mode applies patch message 1`] = `
Object {
  "appRoot": "/",
  "changelog": Object {
    "1.8.6": Object {
      "changes": Object {
        "four": Object {
          "determinedIncrementLevel": "patch",
          "messages": Array [
            Object {
              "affected": Array [
                "four",
              ],
              "body": "",
              "subject": "foo",
              "type": "fix",
            },
          ],
          "version": "4.5.7",
        },
      },
      "date": "Thu, 01 Jan 1970 00:00:00 GMT",
    },
  },
  "config": Object {
    "mode": "range",
  },
  "consts": Object {
    "levels": Array [
      "patch",
      "minor",
      "major",
    ],
    "nsp": "rlsr",
    "rlsrLatest": "rlsr-latest",
  },
  "dbg": [Function],
  "log": [Function],
  "mainPackage": Object {
    "rlsr": Object {
      "determinedIncrementLevel": 0,
    },
    "version": "1.8.6",
  },
  "messages": Array [],
  "packages": Array [
    Object {
      "dependencies": Object {
        "five": "^4.5.0",
        "four": "^4.5.0",
      },
      "name": "one",
      "version": "1.2.3",
    },
    Object {
      "dependencies": Object {
        "three": "2.3.4 - 3",
      },
      "name": "two",
      "version": "2.3.4",
    },
    Object {
      "dependencies": Object {
        "four": "4.0.0 - 4.5.6",
      },
      "name": "three",
      "version": "3.4.5",
    },
    Object {
      "name": "four",
      "version": "4.5.7",
    },
    Object {
      "name": "five",
      "version": "4.5.6",
    },
  ],
}
`;

exports[`large test: exact mode applies several messages 1`] = `
Object {
  "appRoot": "/",
  "changelog": Object {
    "1.9.0": Object {
      "changes": Object {
        "four": Object {
          "determinedIncrementLevel": "minor",
          "messages": Array [
            Object {
              "affected": Array [
                "four",
              ],
              "body": "",
              "subject": "taz",
              "type": "feat",
            },
          ],
          "version": "4.6.0",
        },
        "one": Object {
          "determinedIncrementLevel": "patch",
          "messages": Array [
            Object {
              "affected": Array [
                "one",
              ],
              "body": "",
              "subject": "raz",
              "type": "fix",
            },
          ],
          "version": "1.2.4",
        },
        "three": Object {
          "determinedIncrementLevel": "patch",
          "messages": Array [
            Object {
              "affected": Array [
                "three",
              ],
              "body": "",
              "subject": "roo",
              "type": "fix",
            },
          ],
          "version": "3.4.6",
        },
      },
      "date": "Thu, 01 Jan 1970 00:00:00 GMT",
    },
  },
  "config": Object {
    "mode": "range",
  },
  "consts": Object {
    "levels": Array [
      "patch",
      "minor",
      "major",
    ],
    "nsp": "rlsr",
    "rlsrLatest": "rlsr-latest",
  },
  "dbg": [Function],
  "log": [Function],
  "mainPackage": Object {
    "rlsr": Object {
      "determinedIncrementLevel": 1,
    },
    "version": "1.9.0",
  },
  "messages": Array [],
  "packages": Array [
    Object {
      "dependencies": Object {
        "five": "^4.5.0",
        "four": "^4.5.0",
      },
      "name": "one",
      "version": "1.2.4",
    },
    Object {
      "dependencies": Object {
        "three": "2.3.4 - 3",
      },
      "name": "two",
      "version": "2.3.4",
    },
    Object {
      "dependencies": Object {
        "four": "4.0.0 - 4.5.6",
      },
      "name": "three",
      "version": "3.4.6",
    },
    Object {
      "name": "four",
      "version": "4.6.0",
    },
    Object {
      "name": "five",
      "version": "4.5.6",
    },
  ],
}
`;
